### PR TITLE
[core] Support encrypted redis connection.

### LINKED ISSUE #27371: [Core] Support `rediss://` - TCP protected external Redis instances
### Description

While external Redis instances are supported, it seems that TCP protected Redis instances (which are called with `rediss://`) are not supported.

In the `RAY_REDIS_ADDRESS` env variable you can't supply `rediss://{IP | HOST}:{PORT}`, as it only expects `{IP | HOST}:{PORT}` and doesn't use the `rediss://` schema part.

This would be helpful, as for Ray HA you should be using Redis HA - which is what cloud providers (Azure Cache for Redis, AWS Elasticache) offer, but they are usually setup with TCP protection (eg `rediss://`).

The need for `rediss://` support has already been mentioned earlier at https://github.com/ray-project/ray/issues/17400#issuecomment-901637069, but it seems it was never acted on.

### Use case

The use case is to to connect to cloud-provided Redis instances (Azure Cache for Redis, AWS Elasticache, etc) - which are typically HA Redis instances, but also require `rediss://` for TCP protected connections.

Currently with Ray you cannot use `rediss://` instances as external storage.

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

This PR support encrypted redis connection. If the address starts with `rediss://`, it'll start the redis in SSL mode.

To support better ssl, redis is also upgraded. It's not supposed to impact ray because it's the user's responsibility to provide a redis.

Right now all flags are provided by the os envs. It's not ideal and we should design a better format when we refactor the db layer.

Besides, right now gRPC is using OS envs for the encrypted connection in ray, just follow this pattern for now.

## Related issue number

Closes https://github.com/ray-project/ray/issues/27371
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
hmmm, it seems the building of redis has some issues. I'm looking into this.

**Comment by iycheng** [comment]
```
sentinel.c:34:10: fatal error: openssl/ssl.h: No such file or directory
   34 | #include "openssl/ssl.h"
      |          ^~~~~~~~~~~~~~~
```
I need to change the bazel rule to build redis.

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -655,6 +655,16 @@ RAY_CONFIG(std::string, TLS_SERVER_CERT, "")
 RAY_CONFIG(std::string, TLS_SERVER_KEY, "")
 RAY_CONFIG(std::string, TLS_CA_CERT, "")
 
+/// Location of Redis TLS credentials
+/// https://github.com/redis/hiredis/blob/c78d0926bf169670d15cfc1214e4f5d21673396b/README.md#hiredis-openssl-wrappers
+RAY_CONFIG(bool, REDIS_ENABLE_SSL, false)
```
Comment (scv119) [inline src/ray/common/ray_config_def.h:660]: hmm why upper case?

File: src/ray/gcs/redis_context.cc [hunk 2]
```diff
@@ -255,6 +256,48 @@ void RedisCallbackManager::RemoveCallback(int64_t callback_index) {
...
+    ssl_context_ = redisCreateSSLContext(
+        cacert, capath, client_cert, client_key, server_name, &ssl_error);
+
+    RAY_CHECK(ssl_context_ != nullptr && ssl_error == REDIS_SSL_CTX_NONE)
+        << "Failed to construct a ssl context for redis client: "
+        << redisSSLContextGetError(ssl_error);
+  }
+}
+
+
```
Comment (scv119) [inline src/ray/gcs/redis_context.cc:300]: should we call `redisFreeSSLContext` here in the destructor?

**Review by scv119 — COMMENTED** [review]
also, is it possible to add a test to show it works e2e?

**Review by simon-mo — DISMISSED** [review]
I'm a little bit confused by this PR. What's the workflow for:
- use SSL with the redis bundled/started by Ray (I assume this is supported by the `services.py:_start_redis_instance` change. Do users need to supply all the CERTs parameters?
- Which env var does the user use to connect to hosted redis (e.g. aws memorystore https://docs.aws.amazon.com/memorydb/latest/devguide/getting-startedclusters.connecttonode.html)

**Comment by iycheng** [comment]
@scv119 ah,, I forget to include the unit test. I actually have one but forgot to add.

**Comment by iycheng** [comment]
@simon-mo the services change is only for testing. Ray will not start redis for the user anymore. Ideally, the Redis change should be moved from the service.py to test.py.
So:

>use SSL with the redis bundled/started by Ray (I assume this is supported by the services.py:_start_redis_instance change. Do users need to supply all the CERTs parameters?

It's the user's responsibility to bring a redis. How to setup the cert it depends on the Redis doc.

> Which env var does the user use to connect to hosted redis (e.g. aws memorystore https://docs.aws.amazon.com/memorydb/latest/devguide/getting-startedclusters.connecttonode.html)

I'm not quite sure about AWS's use case. I think here it's more related to the Redis client itself. Everything is based on Redis protocol.

**Comment by iycheng** [comment]
> didn't mean to approve, this PR needs documentation change as well

@simon-mo  right now, all the GCS Fault Tolerance features are only to support ray serve and the only place where ray serves can be HA will be in KubeRay Ray Service Operator. The users can't simply use these features by themselves.

I think we should just add the flags into the kuberay's doc. So it shouldn't be in this one. I'll add it once it's merged.

File: bazel/ray_deps_setup.bzl [hunk 3]
```diff
@@ -98,8 +98,8 @@ def ray_deps_setup():
     auto_http_archive(
         name = "com_github_antirez_redis",
         build_file = "@com_github_ray_project_ray//bazel:BUILD.redis",
-        url = "https://github.com/redis/redis/archive/6.0.10.tar.gz",
-        sha256 = "900cb82227bac58242c9b7668e7113cd952253b256fe04bbdab1b78979cf255a",
+        url = "https://github.com/redis/redis/archive/refs/tags/7.0.4.tar.gz",
```
Comment (ijrsvt) [inline bazel/ray_deps_setup.bzl:101]: Can we switch to `7.0.5` for Security Fixes?

**Review by simon-mo — APPROVED** [review]
can we consider a better API like `RAY_REDIS_ADDRESS=rediss://...`?

**Comment by iycheng** [comment]
@simon-mo  updated to rediss://

**Comment by iycheng** [comment]
test failure not related.

**Pull request merged**